plugins/cuda/cuda_plugin.c

```diff
@@ -0,0 +1,380 @@
+#include "util.h"
+#include "cr_options.h"
+#include "pid.h"
+#include "proc_parse.h"
+
+#include <common/list.h>
+#include <compel/infect.h>
+
+#include <ctype.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <sys/ptrace.h>
+#include <sys/wait.h>
+
+/* cuda-checkpoint binary should live in your PATH */
+#define CUDA_CHECKPOINT "cuda-checkpoint"
+
+/* cuda-checkpoint --action flags */
+#define ACTION_LOCK	  "lock"
+#define ACTION_CHECKPOINT "checkpoint"
+#define ACTION_RESTORE	  "restore"
+#define ACTION_UNLOCK	  "unlock"
+
+#define CUDA_CKPT_BUF_SIZE (128)
+
+#ifdef LOG_PREFIX
+#undef LOG_PREFIX
+#endif
+#define LOG_PREFIX "cuda_plugin: "
+
+struct pid_info {
+	int pid;
+	char checkpointed;
+	struct list_head list;
+};
+
+/* Used to track which PID's we've paused CUDA operations on so far so we can
+ * release them after we're done with the DUMP
+ */
+struct list_head cuda_pids;
+
+static void dealloc_pid_buffer(struct list_head *pid_buf)
+{
+	struct pid_info *info;
+	struct pid_info *n;
+
+	list_for_each_entry_safe(info, n, pid_buf, list) {
+		list_del(&info->list);
+		xfree(info);
+	}
+}
+
+static int add_pid_to_buf(struct list_head *pid_buf, int pid)
+{
+	struct pid_info *new = xmalloc(sizeof(*new));
+
+	if (new == NULL) {
+		return -1;
+	}
+
+	new->pid = pid;
+	new->checkpointed = 0;
+	list_add_tail(&new->list, pid_buf);
+
+	return 0;
+}
+
+static int update_checkpointed_pid(struct list_head *pid_buf, int pid)
+{
+	struct pid_info *info;
+
+	list_for_each_entry(info, pid_buf, list) {
+		if (info->pid == pid) {
+			info->checkpointed = 1;
+			return 0;
+		}
+	}
+
+	return -1;
+}
+
+static int launch_cuda_checkpoint(const char **args, char *buf, int buf_size)
+{
+#define READ  0
+#define WRITE 1
+	int fd[2];
+
+	if (pipe(fd) != 0) {
+		pr_err("Couldn't create pipes for reading cuda-checkpoint output\n");
+		return -1;
+	}
+
+	buf[0] = '\0';
+
+	int child_pid = fork();
+	if (child_pid == -1) {
+		close(fd[READ]);
+		close(fd[WRITE]);
+		pr_err("Failed to fork to exec cuda-checkpoint\n");
+		return -1;
+	}
+
+	if (child_pid == 0) { // child
+		if (dup2(fd[WRITE], STDOUT_FILENO) == -1) {
+			return -1;
+		}
+		if (dup2(fd[WRITE], STDERR_FILENO) == -1) {
+			return -1;
+		}
+		close(fd[READ]);
+		return execvp(args[0], (char **)args);
+	} else { // parent
+		close(fd[WRITE]);
+
+		int bytes_read = read(fd[READ], buf, buf_size);
+		if (bytes_read > 0) {
+			buf[bytes_read - 1] = '\0';
+		}
+
+		int status;
+		waitpid(child_pid, &status, 0);
+		if (!WIFEXITED(status)) {
+			close(fd[READ]);
+			return -1;
+		}
+
+		close(fd[READ]);
+
+		return WEXITSTATUS(status);
+	}
+}
+
+static bool cuda_checkpoint_supports_flag(const char *flag)
+{
+	char msg_buf[2048];
+	const char *args[] = { CUDA_CHECKPOINT, "-h", NULL };
+	int ret = launch_cuda_checkpoint(args, msg_buf, sizeof(msg_buf));
+	if (ret != 0) {
+		pr_err("Failed to launch cuda-checkpoint utility, check that the utility is present in your $PATH\n");
+		return false;
```

Comment:
Instead of failing with an error during `cuda_plugin_init()` when `cuda-checkpoint` is missing, would it be possible to disable the CUDA plugin functionality?